clang/lib/AST/PreorderAST.cpp

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements methods to create and manipulate an n-ary preorder
+//  abstract syntax tree which is used to semantically compare two expressions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/AST/PreorderAST.h"
+
+using namespace clang;
+
+bool PreorderAST::IsDeclOperand(Expr *E, DeclRefExpr *&D) {
+  if (auto *CE = dyn_cast_or_null<CastExpr>(E)) {
+    assert(CE->getSubExpr() && "invalid CastExpr expression");
+
+    if (CE->getCastKind() == CastKind::CK_LValueToRValue ||
+        CE->getCastKind() == CastKind::CK_ArrayToPointerDecay) {
+      E = Lex.IgnoreValuePreservingOperations(Ctx, CE->getSubExpr());
+      if (auto *DRE = dyn_cast_or_null<DeclRefExpr>(E)) {
+        D = DRE;
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void PreorderAST::Create(ASTNode *N, Expr *E, ASTNode *Parent) {
+  if (!E)
+    return;
+
+  // When we invoke Create(N->Left, ...) or Create(N->Right, ...) we need to
+  // create the left or the right nodes with N as the parent node.
+  if (!N)
+    N = new ASTNode(Ctx, Parent);
+
+  // If the parent is non-null, make sure that the current node is marked as a
+  // child of the parent. As a convention, we create left children first.
+  if (Parent) {
+    if (!Parent->Left)
+      Parent->Left = N;
+    else
+      Parent->Right = N;
+  }
+
+  E = Lex.IgnoreValuePreservingOperations(Ctx, E);
+
+  // If E is a variable, store its name in the variable list for the current
+  // node. Initialize the count of the variable to 1.
+  DeclRefExpr *D;
+  if (IsDeclOperand(E, D)) {
+    if (const auto *V = dyn_cast_or_null<VarDecl>(D->getDecl())) {
+      N->AddVar(V->getQualifiedNameAsString());
+      return;
+    }
+  }
+
+  // If E is a constant, store it in the constant field of the current node and
+  // mark that this node has a constant.
+  llvm::APSInt IntVal;
+  if (E->isIntegerConstantExpr(IntVal, Ctx)) {
+    N->Constant = IntVal;
+    N->HasConstant = true;
+    return;
+  }
+
+  if (const auto *BO = dyn_cast<BinaryOperator>(E)) {
+    OpcodeTy Opc = BO->getOpcode();
+    Expr *LHS = BO->getLHS()->IgnoreParens();
+    Expr *RHS = BO->getRHS()->IgnoreParens();
+
+    // Set the Opcode for the current node.
+    N->Opcode = Opc;
+
+    if (isa<BinaryOperator>(LHS))
+      // Create the LHS as the left child of the current node.
+      Create(N->Left, LHS, /* parent node */ N);
+    else
+      // Create the LHS in the current node.
+      Create(N, LHS);
+
+    if (isa<BinaryOperator>(RHS))
+      // Create the RHS as the right child of the current node.
+      Create(N->Right, RHS, /* parent node */ N);
+    else
+      // Create the RHS in the current node.
+      Create(N, RHS);
+  }
+}
+
+void PreorderAST::Sort(ASTNode *N, bool &Error) {
```

Comment:
I find it somewhat confusing that the out parameter Error is named the same as the PreorderAST member Error - is the member PreorderAST::Error ever set during Sort or Normalize, or only the out parameter Error?